# BAN436 Final project

This is a final project for the NHH course BAN436. In this project we will prepare for the Ames Housing Data competition. We will analyze housing data and use predictive models to predict housing prics using simple linear regression.

## Task 1: Import and explore the data

In [26]:
# import all necessary libraries
import pandas as pd
import numpy as np
import plotly.express as px
from sklearn import linear_model

Now that we have all necessary libraries for this project, let's import the data.

In [5]:
# import the data
data = pd.read_csv('AmesHousing.csv')

data

ID  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0     526301100          20       RL  ...       WD          Normal    215000
1     526350040          20       RH  ...       WD          Normal    105000
2     526351010          20       RL  ...       WD          Normal    172000
3     526353030          20       RL  ...       WD          Normal    244000
4     527105010          60       RL  ...       WD          Normal    189900
...         ...         ...      ...  ...       ...            ...       ...
2925  923275080          80       RL  ...       WD          Normal    142500
2926  923276100          20       RL  ...       WD          Normal    131000
2927  923400125          85       RL  ...       WD          Normal    132000
2928  924100070          20       RL  ...       WD          Normal    170000
2929  924151050          60       RL  ...       WD          Normal    188000

[2930 rows x 81 columns]

There seem to be 2930 observations and 81 variables. The first variable is the house ID and last is selling price, so that leaves us with 79 potential explanatory variables. Let's get some more info about the data set.

In [11]:
# get descriptive statistics
data['SalePrice'].describe()

count      2930.000000
mean     180796.060068
std       79886.692357
min       12789.000000
25%      129500.000000
50%      160000.000000
75%      213500.000000
max      755000.000000
Name: SalePrice, dtype: float64

In [10]:
# get info about the variables
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ID             2930 non-null   int64  
 1   MSSubClass     2930 non-null   int64  
 2   MSZoning       2930 non-null   object 
 3   LotFrontage    2440 non-null   float64
 4   LotArea        2930 non-null   int64  
 5   Street         2930 non-null   object 
 6   Alley          198 non-null    object 
 7   LotShape       2930 non-null   object 
 8   LandContour    2930 non-null   object 
 9   Utilities      2930 non-null   object 
 10  LotConfig      2930 non-null   object 
 11  LandSlope      2930 non-null   object 
 12  Neighborhood   2930 non-null   object 
 13  Condition1     2930 non-null   object 
 14  Condition2     2930 non-null   object 
 15  BldgType       2930 non-null   object 
 16  HouseStyle     2930 non-null   object 
 17  OverallQual    2930 non-null   int64  
 18  OverallC

Given the amount of variables, we focus on the sale price to keep it simple. The average price is $180,796.06. The cheapes house was sold for $12,789 and the most expensive one for $755,000. it is worth noting that there is quite a gap between these extremes and lower and upper quartiles. We can assume the presence of a few outliers on both sides.

As for the variables themselves, we notice a majority of categorical variables (43). 36 variables are numerical, some integers other floats. From the data description available we need to also consider some categorical variables are perceived as numerical. We can also already see that there are plenty of missing values in the dataset. For example 'Alley' has only 198 non-null observations out of 2930. This is not because the data is missing but because there is no alley access. Let's look at missing values.

In [68]:
# count missing values
na_values = data.isnull().sum()[data.isnull().sum() > 0]

na_values

LotFrontage      490
Alley           2732
MasVnrType      1775
MasVnrArea        23
BsmtQual          80
BsmtCond          80
BsmtExposure      83
BsmtFinType1      80
BsmtFinSF1         1
BsmtFinType2      81
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Electrical         1
BsmtFullBath       2
BsmtHalfBath       2
FireplaceQu     1422
GarageType       157
GarageYrBlt      159
GarageFinish     159
GarageCars         1
GarageArea         1
GarageQual       159
GarageCond       159
PoolQC          2917
Fence           2358
MiscFeature     2824
dtype: int64

We see many variables have missing data. However, a lot of them are not actually missing, rather they cannot be measured because the feature does not exist. We can see there are approximately 80 homes without basement, 160 homes without garage and 1422 homes without a fireplace. Many are also without pools and fences. Since there is a lot of "missing" data regarding these variables, we cannot simply drop the observations. We will have to assign a value to the instead.

Since we will be predicting home prices with simple regression, there is no need to keep all variables. Let's look at the correlation between sales price and other variables. We can drop those taht have no or very weak correlation later.

In [25]:
# get correlation
data.corr(numeric_only=True)

ID  MSSubClass  LotFrontage  ...    MoSold    YrSold  SalePrice
ID             1.000000   -0.001281    -0.096918  ... -0.050455  0.009579  -0.246521
MSSubClass    -0.001281    1.000000    -0.420135  ...  0.000350 -0.017905  -0.085092
LotFrontage   -0.096918   -0.420135     1.000000  ...  0.011085 -0.007547   0.357318
LotArea        0.034868   -0.204613     0.491313  ...  0.003859 -0.023085   0.266549
OverallQual   -0.263147    0.039419     0.212042  ...  0.031103 -0.020719   0.799262
OverallCond    0.104451   -0.067349    -0.074448  ... -0.007295  0.031207  -0.101697
YearBuilt     -0.343388    0.036579     0.121562  ...  0.014577 -0.013197   0.558426
YearRemodAdd  -0.157111    0.043397     0.091712  ...  0.018048  0.032652   0.532974
MasVnrArea    -0.229283    0.002730     0.222407  ... -0.000276 -0.017715   0.508285
BsmtFinSF1    -0.098375   -0.060075     0.215583  ... -0.001155  0.022397   0.432914
BsmtFinSF2    -0.001145   -0.070946     0.045999  ... -0.009484  0.007105   0.005891
BsmtUnfSF     -0.087707   -0.130421     0.116743  ...  0.021569 -0.036384   0.182855
TotalBsmtSF   -0.189642   -0.219445     0.353773  ...  0.016678 -0.010405   0.632280
1stFlrSF      -0.141902   -0.247828     0.457391  ...  0.040496 -0.013667   0.621676
2ndFlrSF      -0.003289    0.304237     0.029187  ...  0.013247 -0.018530   0.269373
LowQualFinSF   0.056940    0.025765     0.005249  ...  0.011397 -0.002074  -0.037660
GrLivArea     -0.107579    0.068061     0.383822  ...  0.043665 -0.026489   0.706780
BsmtFullBath  -0.037759    0.013701     0.108915  ... -0.003471  0.044905   0.276050
BsmtHalfBath   0.004328   -0.003329    -0.024724  ...  0.022699 -0.019529  -0.035835
FullBath      -0.171431    0.134631     0.184521  ...  0.046032 -0.004754   0.545604
HalfBath      -0.166636    0.175879     0.041880  ... -0.001311  0.001561   0.285056
BedroomAbvGr   0.006345   -0.019208     0.240442  ...  0.053677 -0.018008   0.143913
KitchenAbvGr   0.076470    0.257698     0.005407  ...  0.035201  0.035421  -0.119814
TotRmsAbvGrd  -0.068981    0.031898     0.353137  ...  0.043784 -0.030498   0.495474
Fireplaces    -0.108056   -0.049955     0.257255  ...  0.032152 -0.007612   0.474558
GarageYrBlt   -0.256829    0.088754     0.076306  ...  0.024498 -0.005159   0.526965
GarageCars    -0.237484   -0.045883     0.308706  ...  0.049847 -0.022488   0.647877
GarageArea    -0.210606   -0.103239     0.358505  ...  0.039544 -0.013018   0.640401
WoodDeckSF    -0.051135   -0.017310     0.120084  ...  0.016974  0.000882   0.327143
OpenPorchSF   -0.071311   -0.014823     0.163040  ...  0.033651 -0.037467   0.312951
EnclosedPorch  0.162519   -0.022866     0.012758  ... -0.021324 -0.000505  -0.128787
3SsnPorch     -0.024894   -0.037956     0.028564  ...  0.027229  0.022668   0.032225
ScreenPorch   -0.025735   -0.050614     0.076666  ...  0.028169 -0.006116   0.112151
PoolArea      -0.002845   -0.003434     0.173947  ... -0.042223 -0.052541   0.068403
MiscVal       -0.008260   -0.029254     0.044476  ...  0.007333  0.008574  -0.015691
MoSold        -0.050455    0.000350     0.011085  ...  1.000000 -0.155554   0.035259
YrSold         0.009579   -0.017905    -0.007547  ... -0.155554  1.000000  -0.030569
SalePrice     -0.246521   -0.085092     0.357318  ...  0.035259 -0.030569   1.000000

[38 rows x 38 columns]

From a quick analysis we can see that there are quite high correlations between sales price and different variables. The highest are for example overall quality, year of construction or renovation, area of different features. Note that these are only numeric variables. We will transform the data later and choose the variables then.

## Task 2: Visualize the data

We can also analyze tha data visually. Let's create a few plots to understand possible relationships of the data. Firs, let's look at the dependent variable 'SalePrice'.

In [28]:
# histogram
plot_1 = px.histogram(
    data,
    x='SalePrice'
)

plot_1.show()

We can see that there are noticable outliers of homes that cost more that $500k. These observations might affect results significnatly so it is worht considering droping these.

Next, let's consider the correlations. We noticed that there is a strong correlation between sales price and other variables. Scatterplots should give us a better idea what kind of relationship it is (linear or other). We will use a function so we can choose any variables anytime.

In [31]:
# scatterplot function
def print_scatterplots(data, dep, expl):
    for var in expl:
        plot = px.scatter(
            data,
            x=var,
            y=dep,
            title=f'{dep} X {var}'
        )

        plot.show()

In [32]:
print_scatterplots(data, 'SalePrice', ['OverallQual', 'YearBuilt', 'YearRemodAdd', '1stFlrSF'])

We can see possible linear relationships between sales price and other explanatory variables. We can reuse the function with other variables if necessary.

## Task 3: Data wrangling

Now, the time has come to address the flaws of the data. We need to handle the missing values, transform string data to categorical variables, remove unwanted outliers and decide on the variables we will use for predictions.

#### Missing values

There are 27 variables with missing values. We need to identify truly missing values and data unavailable for simply not existing. Let's assume that missing values in numeric variables are labeled correctly as missing. We will get rid of these observations as there should not be that many of them. This also allows us to keep the same observations for each simple regression, it will be easier to compare the results.

In [77]:
# get a copy of the data to clean
data_clean = data.copy()

# filter and delete missing values from numeric variables
data_clean = data_clean.dropna(subset=data_clean[na_values.index].select_dtypes('float').columns)

na_values_clean = data_clean.isnull().sum()[data_clean.isnull().sum() > 0]

data_clean[na_values_clean.index].info()

<class 'pandas.core.frame.DataFrame'>
Index: 2274 entries, 0 to 2929
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Alley         162 non-null    object
 1   MasVnrType    913 non-null    object
 2   BsmtQual      2223 non-null   object
 3   BsmtCond      2223 non-null   object
 4   BsmtExposure  2220 non-null   object
 5   BsmtFinType1  2223 non-null   object
 6   BsmtFinType2  2222 non-null   object
 7   Electrical    2273 non-null   object
 8   FireplaceQu   1174 non-null   object
 9   PoolQC        12 non-null     object
 10  Fence         445 non-null    object
 11  MiscFeature   74 non-null     object
dtypes: object(12)
memory usage: 231.0+ KB


Now we have data where missing values are only for string (object) type variables. Notice that the code before also dropped 4 variables that had the same missing observations. Lucky us now have to deal with only 12 variables. After consulting the data description file, we see that except "Electrical" variable, all other are simply missing the features, not observations. Lets drop the one observation for "Electrical" and replace the NA values of other variables with "none".

In [81]:
# drop missing values for "electrical" variable
data_clean = data_clean.dropna(subset='Electrical')

# replace other missing values with "none"
data_clean = data_clean.fillna('none')

data_clean.isna().sum() # no missing values

ID               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
SalePrice        0
Length: 81, dtype: int64

#### Categorical variables

In this section we will need to transform and recode some of the variables. There are four types of variables we need to deal with:
* Categorical as numeric (ordered)
* Categorical as numeric (unordered)
* Categorical as string (ordered)
* Categorical as string (unordered)

With pandas, it is easy to transform data into categorical type


In [91]:
df = data_clean.copy()

# df['MSSubClass'] = df['MSSubClass'].astype('category')
df['MSSubClass'] = pd.Categorical(df['MSSubClass'], ordered=True)

df['MSSubClass'].max()

np.int64(190)